### PR TITLE
feat: recursive jar unpacking by desired depth

### PR DESCRIPTION
**Pull request by shaninja**
- [x] Ready for review
- [x] Follows CONTRIBUTING rules
- [ ] Reviewed by Snyk internal team

#### What does this PR do?
Up until now, we only unpacked 1 level of jars, given `shaded-jars` flag was provided.

This PR changes the flag to be `shaded-jars-depth`, which accept as a number the number of levels of jars the user wants us to unpack.

#### Any background context you want to provide?
Customers that would use this are aware of performance implications.

In the future, we will log in the CLI the number of levels we have managed to unpack, so the user can gain confidence if they got a full scan.

#### What are the relevant tickets?
CAP-412

#### Screenshots
(deleted all for now)
**no flag**
test:

monitor:

**unsupported type**

**no --app-vulns**

**--shaded-jars-depth=0** (should be same as no flag)
test:

monitor:

**no number given** (should be same as =1)
test:

monitor:

**--shaded-jars-depth=1** 
test:

monitor: BBBAAADD??

**--shaded-jars-depth=2**
test:

monitor:

**--shaded-jars-depth=3** 
test:

monitor:

### CODE REVIEW

File: lib/analyzer/applications/java.ts [hunk 1]
```diff
@@ -80,40 +96,90 @@ function getJarSha(jarBuffers: JarBuffer[]): JarFingerprint[] {
...
-      fingerprints.push({
-        location: jarBuffer.location,
-        digest: bufferToSha1(jarBuffer.digest),
+    if (foundJars === false) {
+      jarBuffers.push({
+        location: jarPath,
+        digest: jarBuffer,
       });
+    } else {
+      // if we found jars, we unpacked them (called unpackJarsTraverse again)
```
Comment (jurglic) [inline lib/analyzer/applications/java.ts:151]: So parent jars (that contain other jars) are never included in results? I guess these are fat/uber jars and we will never track them as such?

If I'm getting this correctly, then perhaps we could remove dead branch and rather comment line 145, that `!foundJars` means that current jar is not fat, therefore we include it?

File: lib/analyzer/applications/java.ts [hunk 2]
```diff
@@ -80,40 +96,90 @@ function getJarSha(jarBuffers: JarBuffer[]): JarFingerprint[] {
...
     }
   }
 
+  return jarBuffers;
+}
+
+function checkIfFatJarsAndUnpack(
```
Comment (jurglic) [inline lib/analyzer/applications/java.ts:158]: Nitpick, maybe just `unpackFatJars`?

**Review by jurglic — APPROVED** [review]
Had a look and the code is pretty clear! Just added couple of nitpick comments.. 😄

File: lib/analyzer/applications/java.ts (hunk 1 above)
Comment (agatakrajewska) [reply lib/analyzer/applications/java.ts:151]: > So parent jars (that contain other jars) are never included in results? I guess these are fat/uber jars and we will never track them as such?

That is correct, uber jar is essentially a wrapper so we don't care, we know that those hashes won't return anything from maven central, but we will see them in the path, later in registry 

> If I'm getting this correctly, then perhaps we could remove dead branch and rather comment line 145, that !foundJars means that current jar is not fat, therefore we include it?

That is nicer yes, I'll update!

File: lib/analyzer/applications/java.ts [hunk 3]
```diff
@@ -80,40 +96,90 @@ function getJarSha(jarBuffers: JarBuffer[]): JarFingerprint[] {
...
+    });
+
+    fingerprints.push(
+      ...jars.map((element) => {
+        return {
+          ...element,
+          digest: bufferToSha1(element.digest),
+        };
+      }),
+    );
```
[the remark this answers is not in the document]
Comment (jurglic) [reply lib/analyzer/applications/java.ts:173]: Something like `fingerprints.push(...getJarShas(jars))`?

**Review by jurglic — APPROVED** [review]

File: lib/analyzer/applications/java.ts [hunk 4]
```diff
@@ -71,6 +71,22 @@ export async function jarFilesToScannedProjects(
...
+) {
+  let fingerprints;
+  if (desiredLevelsOfUnpacking > 0) {
+    fingerprints = checkIfFatJarsAndUnpack(
+      jarBuffers,
+      desiredLevelsOfUnpacking,
+    );
+  } else {
+    fingerprints = getJarSha(jarBuffers);
+  }
```
[the remark this answers is not in the document]
Comment (agatakrajewska) [reply lib/analyzer/applications/java.ts:80]: Done, I've refactored it nicely!

File: lib/analyzer/applications/java.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (agatakrajewska) [reply lib/analyzer/applications/java.ts]: Done :-)

**Review by agranado2k — APPROVED** [review]
LGTM! 🎉

**Comment by shaninja** [comment]
thank you @jurglic for all the great suggestions!

**Pull request merged**